planet/src/main/webapp/js/placesItinerary.js

```diff
@@ -52,33 +54,65 @@ function updateStartingAddress() {
 
 export function renderPlaces() {
     displayedPlaces = {};
+    toBeProcessedPlaces = [];
     const userId = firebase.auth().currentUser.uid;
     const placesRef = database.ref('users/' + userId + '/places');
-    placesRef.once('value', (placesSnapshot) => {
+    placesRef.orderByValue().once('value', (placesSnapshot) => {
         const placesContainer = document.getElementById('places');
         placesContainer.innerHTML = '';
-        const savedPlaces = placesSnapshot.val();
-        for (const placeId in savedPlaces) {
-            // Make request with fields
-            let placeRequest = {
-                placeId: placeId,
-                fields: ['place_id','name','rating','formatted_address',
-                    'opening_hours','photos','url']
-            };
-            let service = new google.maps.places.PlacesService(map);
-            service.getDetails(placeRequest, renderPlaceDetailsCallback); 
-        }
+        placesSnapshot.forEach(function(childPlace) {
+            const placeId = childPlace.key;
+            toBeProcessedPlaces.push(placeId);
+        });
+        makePlaceRequests();
     });
 }
 
+function makePlaceRequests() {
+    hideLoadMoreButton();
+    let service = new google.maps.places.PlacesService(map);
+    const maxNumOfRequests = 10;
+    for (let i = 0; i < Math.min(maxNumOfRequests, toBeProcessedPlaces.length); i++) {
+        let placeId = toBeProcessedPlaces[i];
+        // Make request with fields
+        let placeRequest = {
+            placeId: placeId,
+            fields: ['place_id','name','rating','formatted_address',
+                'opening_hours','photos','url']
+        };
+        service.getDetails(placeRequest, renderPlaceDetailsCallback);
+    } 
+    // Google Maps API can only process 10 place details requests at a time,
+    // and once the session is full, new requests are refilled at a rate of 2 per second.
+    // Therefore, we need to wait 5 seconds here to process another group of 10 requests.
+    setTimeout(showLoadMoreButton, 5000);
+}
+
 function renderPlaceDetailsCallback(place, status) {
     if (status == google.maps.places.PlacesServiceStatus.OK) {
         const placesContainer = document.getElementById('places');
         const placeElement = createPlaceElement(place);
         placesContainer.appendChild(placeElement);
+
+        const indexOfProcessedPlace = toBeProcessedPlaces.indexOf(place.place_id);
+        if (indexOfProcessedPlace > -1) {
+            toBeProcessedPlaces.splice(indexOfProcessedPlace, 1);
+        }
```

Comment:
nit: using a list makes this code a bit awkward. Consider using a dict instead (in JS that's an object, I think)

Reply:
True. But using an object means we have to keep a counter when looping the first 10 items, which might be slightly awkward too...

Reply:
I'm not sure I follow why that would be necessary. Using an object for toBeProcessedPlaces would allow you to remove with an easier to understand piece of code:

`delete toBeProcessedPlaces[place.place_id]`

If I'm misunderstanding, feel free to explain here or schedule a quick call for us to discuss

Reply:
haha I was talking about when we need to send another batch of 10 requests, if `toBeProcessedPlaces` is an object, we will need to call `object.keys()` to get an array of placeIds, so that we can loop through the first 10 keys (or a forEach loop with a counter). We'll also need to do the same thing for getting the object size (`object.keys().length`)
The delete function will be easier, but looping the first 10 items will be more complicated

Reply:
ok it seems like this is a reasonable tradeoff.